nixos/modules/services/web-apps/mobilizon.nix

```diff
@@ -0,0 +1,358 @@
+in
+{
+  options = {
+    services.mobilizon = {
+      enable = mkEnableOption
+        "Mobilizon federated organization and mobilization platform";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.mobilizon;
+        description = ''
+          Which Mobilizon package to use.
+        '';
+      };
+
+      settings = mkOption {
+        type =
+          let
+            elixirTypes = settingsFormat.lib.types;
+          in
+          types.submodule {
+            freeformType = settingsFormat.type;
+
+            options = {
+              ":mobilizon" = {
+
+                "Mobilizon.Web.Endpoint" = {
+                  url.host = mkOption {
+                    type = elixirTypes.str;
+                    defaultText = literalExample ''
+                      ''${settings.":mobilizon".":instance".hostname}
+                    '';
+                    description = ''
+                      Your instance's hostname for generating URLs throughout the app
+                    '';
+                  };
+
+                  http.port = mkOption {
+                    type = elixirTypes.port;
+                    default = 4000;
+                    description = ''
+                      The port to run the server
+                    '';
+                  };
+
+                  has_reverse_proxy = mkOption {
+                    type = elixirTypes.bool;
+                    default = true;
+                    description = ''
+                      Whether you use a reverse proxy
+                    '';
+                  };
+                };
+
+                ":instance" = {
+                  name = mkOption {
+                    type = elixirTypes.str;
+                    description = ''
+                      The fallback instance name if not configured into the admin UI
+                    '';
+                  };
+
+                  hostname = mkOption {
+                    type = elixirTypes.str;
+                    description = ''
+                      Your instance's hostname
+                    '';
+                  };
+
+                  email_from = mkOption {
+                    type = elixirTypes.str;
+                    defaultText = literalExample ''
+                      noreply@''${settings.":mobilizon".":instance".hostname}
+                    '';
+                    description = ''
+                      The email for the From: header in emails
+                    '';
+                  };
+
+                  email_reply_to = mkOption {
+                    type = elixirTypes.str;
+                    defaultText = literalExample ''
+                      ''${email_from}
+                    '';
+                    description = ''
+                      The email for the Reply-To: header in emails
+                    '';
+                  };
+                };
+
+                "Mobilizon.Storage.Repo" = {
```

Comment:
One question regarding this.
Here you are using the postgres.ensureUsers to make sure the permission are granted to the user, so that you don't have to connect to the database with a password. Am I right?
Essentially you can use a DB_URL like `user:@host:port/db` (i.e. without the `password` part)
I'm currently trying to thing about how to configure elixir services and the need for environmentFiles. That's why I'm asking.

Reply:
I think the DB_URL would make things more complicated for me, since I'd have to parse it, and differentiate between connecting to Postgres using the UNIX socket, or TCP, and then configure Mobilizon using that, instead of asking the user to look up upstream's documentation, and use the configuration directly. This is also why I didn't configure Mobilizon to be a DynamicUser, because Postgres to know of the user beforehand to do proper authentication.

Reply:
I feel like there's also a big disadvantage in using environment files in nixos, it's that it doesn't compose well between multiple services.

For example, if I were to ask the user for the location of an environment file where it has a DB_URL variable, this value is really hard to read from the point of view of other services.

If I have a service that backs up Postgres databases, I would want to configure it to point it to whatever Mobilizon is using. With environment files, I'd have to include it, parse the variable at run time (hoping that there is no collisions), etc.

With RFC42-style configuration, I can just use `config.services.mobilizon.":mobilizon"."Mobilizon.Storage.Repo"` and do what I want at evaluation time, without any need of parsing.

I think using environment files for secrets is fine though, as you rarely need to compose services on secrets.

Reply:
I was not so clear here.
I agree with your approach and don't advise on using DB_URL. I'm just wondering about how the magic happens that the app can connect to the database.
If I remember correctly you need to set the authentication variable of the postgres db in a special way.
I'm trying to get all the informations out so they can be in the documentation. Maybe I'm missing something.
Sorry to bother you with those details.
You are the first one to go through an elixir PR, once we figure out this stuff, the following ones should be much faster!

I'm currently making experiments with SnvironmentFiles. the systemd doc says you are not supposed to use that for secrets. it talks about LoadCredentials. I haven't found any nix examples though. I'm digging through that stuff. I'm really starting to appreciate your approach of not going through that stuff :-)

Reply:
No need to apologize, I'm more than happy to use this PR as a way of figuring out how to package Elixir software for NixOS.

In this PR, there was no need to set any environment variable for instructing Mobilizon to connect to the database. Mobilizon uses Ecto, through the module `Mobilizon.Storage.Repo`. It therefore adds configuration options under `:mobilizon` -> `Mobilizon.Storage.Repo`, with [these values](https://hexdocs.pm/ecto_sql/Ecto.Adapters.Postgres.html) available. I wanted to connect to Postgres using UNIX socket, so I just had to set a default for the `socket_dir` value. This gets converted, along all other configuration in `settings`, into an Elixir configuration using the code from #119049. If it were any other app, I would have then placed this configuration into the `runtime.exs` file, like we discussed a while ago, but since Mobilizon added a [way](https://framagit.org/framasoft/mobilizon/-/blob/1.1.1/mix.exs#L36) to [load external configuration files](https://framagit.org/framasoft/mobilizon/-/blob/1.1.1/lib/config_provider.ex), I used that instead.

As far as I know, you could configure the database configuration using environment variables here 2 different ways:

- Using variables that are read from the Postgrex library, which is what the Ecto Postgres adapter uses, but to me, the fact that Ecto uses Postgrex is an implementation detail, so we shouldn't rely on that.
- If from anywhere in the configuration, we have a snippet like this:

```elixir
import Config

config :mobilizon, Mobilizon.Storage.Repo,
  username: System.get_env("MOBILIZON_DATABASE_USERNAME", "mobilizon"),
  password: System.get_env("MOBILIZON_DATABASE_PASSWORD", "mobilizon"),
  # ...
```

But upstream this is set only in `dev.exs`, which we're not using, or in `docker.exs` which we could use, but at the end of the day, we have a system with which we can set any configuration option, generically, for any Elixir app, which is more direct than adding an indirection through an environment variable.

I hope this is clears up the magic, and I hope you didn't mind me going into too much detail.

IIRC SetCredential and LoadCredentials are a very recent feature of systemd (>= 247), which is probably why it's not really common in NixOS yet. I have also yet to investigate how it works / how we could use it in NixOS. For reference, a discussion was started here: https://github.com/NixOS/nixpkgs/issues/102397

Reply:
Thanks a lot for going into details here. 
What is still unclear to me is how you can connect without a password.
If I understand correctly, using the socket_dir, you can connect to the socket without a password.
I thought that you had to enable connection without a password in the postgres authentication file for that to work (the pg_hba.conf).

Reply:
It uses a line that is actually already included in most distributions of Postgresql:

```
# type database user address method
# "local" is for Unix domain socket connections only
local  all      all          trust 
```

This allows anyone who has a connection to postgresql using the unix socket in the socket_dir to be already trusted, provided that a postgres user with the same name as the unix user connecting to the socket exists. The security model is that, if someone has already access to the filesystem to that extent, it must already be a trusted process.

Reply:
I had a look at the doc.
https://www.postgresql.org/docs/current/auth-pg-hba-conf.html
trust is unconditional. peer is the one that checks that the user with the same name exists.

Maybe we should add something somewhere in the docs for setting up the database? Not sure if it belongs here, but there is basically no documentation for setting up postgres with nix.

Reply:
One more thing on this, peer or ident authentication (checking that the user name is correct) is not the default.
https://gist.github.com/saveriomiroddi/fc8c4e285cb7296cdf84b4a7df53d31b
(this is old but still valid). The default for TCP/IP connections being md5.
It might be good to write somewhere that their permissions to run this need to be changed.

Reply:
After checking some more on this.
The default authentication for postgres is not peer for localhost connections. Since here you are using socketdir, the default is peer, and so you don't need to write any particular docs.
I was previously using a db url, so going through TCP/IP, and for that the default is that you need a password.
I'll update if i find out anything, but basically, this is good enough and doesn't need any additional documentation in my opinion.

Reply:
Just a little update on this. At the moment there is a PR on plausible which is privacy friendly analytics in elixir.
The person making the PR chose DynamicUser because it creates a user with the same name as the application. So he just ensures the permissions in postgres and doesn't need to create a user.
Just mentioning this as an alternative, not at all as the best way to do things.
I thought it was interesting that it was possible to do it like that.